File: _includes/code/howto/manage-data.import.py
# How-to: Manage-data -> (Batch) Import items - Python examples
import os
MAX_ROWS_TO_IMPORT = 50  # limit vectorization calls

# ================================
# ===== INSTANTIATION-COMMON =====
# ================================

import weaviate

# Instantiate the client with the OpenAI API key
client = weaviate.connect_to_local(
    headers={
        "X-OpenAI-Api-Key": os.environ["OPENAI_API_KEY"]  # Replace with your inference API key
    }
)

assert client.is_ready()

# ============================
# ===== Define the class =====
# ============================

# Clean slate
client.collections.delete("YourCollection")

from weaviate.classes.config import Configure

client.collections.create(
    "YourCollection",
    vectorizer_config=Configure.Vectorizer.none()
)

# ==============================
# ===== Basic batch import =====
# ==============================

# BasicBatchImportExample
data_rows = [
    {"title": f"Object {i+1}"} for i in range(5)
]

collection = client.collections.get("YourCollection")

# highlight-start
with collection.batch.dynamic() as batch:
    for data_row in data_rows:
        batch.add_object(
            properties=data_row,
        )
# highlight-end
# END BasicBatchImportExample

result = collection.aggregate.over_all(total_count=True)
assert result.total_count == 5

# Clean up
client.collections.delete(collection.name)

# =======================================
# ===== Insert_many with custom ID =====
# =======================================

# InsertManyWithIDExample
# highlight-start
from weaviate.util import generate_uuid5  # Generate a deterministic ID
from weaviate.classes.data import DataObject
# highlight-end

data = [
    # use DataObject to provide uuid value
    DataObject(
        properties={"title": "Object 1"},
        # highlight-start
        uuid=generate_uuid5({"title": "Object 1"})
        # highlight-end
    ),
    DataObject(
        properties={"title": "Object 2"},
        uuid=generate_uuid5({"title": "Object 2"})
    ),
    DataObject(
        properties={"title": "Object 3"},
        uuid=generate_uuid5({"title": "Object 3"})
    ),
]

collection = client.collections.get("YourCollection")  # Replace with your collection name
collection.data.insert_many(data)
# END InsertManytWithIDExample

# Tests
result = collection.aggregate.over_all(total_count=True)
assert result.total_count == 3

first_id = generate_uuid5({"title": "Object 1"})
response = collection.query.fetch_object_by_id(first_id)
assert response != None

# Clean up
client.collections.delete(collection.name)

# ===========================================
# ===== Insert many with custom vector =====
# ===========================================

# InsertManyWithVectorExample
from weaviate.classes.data import DataObject

data = [
    # use DataObject to provide custom vector
    DataObject(
        properties={"title": "Object 1"},
        # highlight-start
        vector=[0.1, 0.1, 0.1, 0.1, 0.1, 0.1]
        # highlight-end
    ),
    DataObject(
        properties={"title": "Object 2"},
        vector=[0.2, 0.2, 0.2, 0.2, 0.2, 0.2]
    ),
    DataObject(
        properties={"title": "Object 3"},
        vector=[0.3, 0.3, 0.3, 0.3, 0.3, 0.3]
    ),
]

collection = client.collections.get("YourCollection")  # Replace with your collection name
collection.data.insert_many(data)
# END InsertManyWithVectorExample

# Tests
result = collection.aggregate.over_all(total_count=True)
assert result.total_count == 3

response = collection.query.bm25(
    query="Object 1",
    include_vector=True
)
test_vector = response.objects[0].vector["default"]
assert (test_vector[0] >= 0.1)
assert (test_vector[0] < 0.11)

# Clean up
client.collections.delete(collection.name)


# =======================================
# ===== Insert many with custom ID =====
# =======================================

# BatchImportWithIDExample
# highlight-start
from weaviate.util import generate_uuid5  # Generate a deterministic ID
# highlight-end

# BatchImportWithIDExample
data_rows = [{"title": f"Object {i+1}"} for i in range(5)]

collection = client.collections.get("YourCollection")

# highlight-start
with collection.batch.dynamic() as batch:
    for data_row in data_rows:
        obj_uuid = generate_uuid5(data_row)
        batch.add_object(
            properties=data_row,
            uuid=obj_uuid
        )
# highlight-end
# END BatchImportWithIDExample

result = collection.aggregate.over_all(total_count=True)
assert result.total_count == 5
resp_obj = collection.query.fetch_object_by_id(obj_uuid)
assert resp_obj != None
# Clean up
client.collections.delete(collection.name)

# ===========================================
# ===== Batch import with custom vector =====
# ===========================================

# BatchImportWithVectorExample
data_rows = [{"title": f"Object {i+1}"} for i in range(5)]
vectors = [[0.1] * 1536 for i in range(5)]

collection = client.collections.get("YourCollection")

# highlight-start
with collection.batch.dynamic() as batch:
    for i, data_row in enumerate(data_rows):
        batch.add_object(
            properties=data_row,
            vector=vectors[i]
        )
# highlight-end
# END BatchImportWithVectorExample

result = collection.aggregate.over_all(total_count=True)
assert result.total_count == 5
# Clean up
client.collections.delete(collection.name)


# ===========================================
# ===== Batch import with named vectors =====
# ===========================================

# Clean slate
client.collections.delete("YourCollection")

from weaviate.classes.config import Configure, Property, DataType

client.collections.create(
    name="YourCollection",
    properties=[
        Property(name="title", data_type=DataType.TEXT),
        Property(name="body", data_type=DataType.TEXT),
    ],
    vectorizer_config=[
        Configure.NamedVectors.text2vec_openai(
            name="title",
            source_properties=["title"]
        ),
        Configure.NamedVectors.text2vec_openai(
            name="body",
            source_properties=["body"]
        ),
    ]
)

# BatchImportWithNamedVectors
data_rows = [{
    "title": f"Object {i+1}",
    "body": f"Body {i+1}"
} for i in range(5)]

title_vectors = [[0.12] * 1536 for _ in range(5)]
body_vectors = [[0.34] * 1536 for _ in range(5)]

collection = client.collections.get("YourCollection")

# highlight-start
with collection.batch.dynamic() as batch:
    for i, data_row in enumerate(data_rows):
        batch.add_object(
            properties=data_row,
            vector={
                "title": title_vectors[i],
                "body": body_vectors[i],
            }
        )
# highlight-end
# END BatchImportWithNamedVectors

response = collection.query.fetch_objects(include_vector=True)
print(len(response.objects))
for i, obj in enumerate(response.objects):
    print(obj.vector)
    assert "Object" in obj.properties["title"]
    assert "Body" in obj.properties["body"]
    assert set(obj.vector.keys()) == {"title", "body"}
    assert len(obj.vector["title"]) == 1536
    assert len(obj.vector["body"]) == 1536


# =======================================
# ===== Batch import with cross-reference =====
# =======================================

client.collections.delete(["Author", "Publication"])

from weaviate.classes.config import Property, ReferenceProperty, DataType

publications = client.collections.create(
    name="Publication",
    properties=[Property(name="title", data_type=DataType.TEXT)]
)

authors = client.collections.create(
    name="Author",
    properties=[Property(name="name", data_type=DataType.TEXT)],
    references=[
        ReferenceProperty(name="writesFor", target_collection="Publication")
    ]
)

from_uuid = authors.data.insert(
    properties={"name": "Jane Austen"}
)

publications.data.insert(
    {"title": "Ye Olde Times"}
)
target_uuid = publications.query.fetch_objects(limit=1).objects[0].uuid

# BatchImportWithRefExample
collection = client.collections.get("Author")

with collection.batch.fixed_size(batch_size=100) as batch:
    batch.add_reference(
        from_property="writesFor",
        from_uuid=from_uuid,
        to=target_uuid,
    )

# END BatchImportWithRefExample

# Tests
from weaviate.classes.query import QueryReference

response = collection.query.fetch_object_by_id(
    from_uuid,
    return_references=QueryReference(link_on="writesFor", return_properties=["title"])
)

assert response.references["writesFor"].objects[0].properties["title"] == "Ye Olde Times"

# Clean up
client.collections.delete(collection.name)


# ===========================================
# ===== Stream data - JSON =====
# ===========================================

# Clean slate
client.collections.delete("JeopardyQuestion")

# Download the data from https://raw.githubusercontent.com/weaviate-tutorials/edu-datasets/main/jeopardy_1k.json
# and save it in the same folder as this script

import requests
import json

request = requests.get("https://raw.githubusercontent.com/weaviate-tutorials/edu-datasets/main/jeopardy_1k.json")
with open("jeopardy_1k.json", "w") as f:
    f.write(request.text)


# START JSON streaming
import ijson

# Settings for displaying the import progress
counter = 0
interval = 100  # print progress every this many records; should be bigger than the batch_size

print("JSON streaming, to avoid running out of memory on large files...")
with client.batch.fixed_size(batch_size=200) as batch:
    with open("jeopardy_1k.json", "rb") as f:
        objects = ijson.items(f, "item")
        for obj in objects:
            properties = {
                "question": obj["Question"],
                "answer": obj["Answer"],
            }
            batch.add_object(
                collection="JeopardyQuestion",
                properties=properties,
                # If you Bring Your Own Vectors, add the `vector` parameter here
                # vector=obj.vector["default"]
            )

            # Calculate and display progress
            counter += 1
            if counter % interval == 0:
                print(f"Imported {counter} articles...")


print(f"Finished importing {counter} articles.")
# END JSON streaming

# Tests
questions = client.collections.get("JeopardyQuestion")
response = questions.aggregate.over_all(total_count=True)
assert response.total_count == 1000

# Cleanup
client.collections.delete("JeopardyQuestion")

# ===========================================
# ===== Stream data - CSV =====
# ===========================================

import pandas as pd

df = pd.read_json("jeopardy_1k.json")
df.to_csv("jeopardy_1k.csv", index=False)

# START CSV streaming
import pandas as pd

# Settings for displaying the import progress
counter = 0
interval = 100  # print progress every this many records; should be bigger than the batch_size

def add_object(obj) -> None:
    global counter
    properties = {
        "question": obj["Question"],
        "answer": obj["Answer"],
    }

    with client.batch.fixed_size(batch_size=200) as batch:
        batch.add_object(
            collection="JeopardyQuestion",
            properties=properties,
            # If you Bring Your Own Vectors, add the `vector` parameter here
            # vector=obj.vector["default"]
        )

        # Calculate and display progress
        counter += 1
        if counter % interval == 0:
            print(f"Imported {counter} articles...")


print("pandas dataframe iterator with lazy-loading, to not load all records in RAM at once...")
with client.batch.fixed_size(batch_size=200) as batch:
    with pd.read_csv(
        "jeopardy_1k.csv",
        usecols=["Question", "Answer", "Category"],
        chunksize=100,  # number of rows per chunk
    ) as csv_iterator:
        # Iterate through the dataframe chunks and add each CSV record to the batch
        for chunk in csv_iterator:
            for index, row in chunk.iterrows():
                properties = {
                    "question": obj["Question"],
                    "answer": obj["Answer"],
                }
                batch.add_object(
                    collection="JeopardyQuestion",
                    properties=properties,
                    # If you Bring Your Own Vectors, add the `vector` parameter here
                    # vector=obj.vector["default"]
                )

        # Calculate and display progress
        counter += 1
        if counter % interval == 0:
            print(f"Imported {counter} articles...")

print(f"Finished importing {counter} articles.")
# END CSV streaming

# Tests
questions = client.collections.get("JeopardyQuestion")
response = questions.aggregate.over_all(total_count=True)
assert response.total_count == 1000

# Cleanup
client.collections.delete("JeopardyQuestion")

# Delete the downloaded files
os.remove("jeopardy_1k.json")
os.remove("jeopardy_1k.csv")

# ================================================
# =====   Batch vectorization set parameters =====
# ================================================

import os
rpm_embeddings = 100
tpm_embeddings = 10000

cohere_key = os.environ["COHERE_API_KEY"]
openai_key = os.environ["OPENAI_API_KEY"]


# START BatchVectorizationClientModify
from weaviate.classes.config import Integrations

integrations = [
    # Each model provider may expose different parameters
    Integrations.cohere(
        api_key=cohere_key,
        requests_per_minute_embeddings=rpm_embeddings,
    ),
    Integrations.openai(
        api_key=openai_key,
        requests_per_minute_embeddings=rpm_embeddings,
        tokens_per_minute_embeddings=tpm_embeddings,   # e.g. OpenAI also exposes tokens per minute for embeddings
    ),
]
client.integrations.configure(integrations)
# END BatchVectorizationClientModify

collection = client.collections.get("NewCollection")

collection.data.insert_many(
    {"title": f"Some title {i}", "summary": f"Summary {i}", "body": f"Body {i}"} for i in range(5)
)

response = collection.aggregate.over_all(total_count=True)
assert response.total_count == 5

# Clean up
client.collections.delete("NewCollection")

client.close()

File: _includes/code/howto/manage-data.import.ts
// How-to: Manage data -> (Batch) Import items - TypeScript examples
// run with: node --loader=ts-node/esm --experimental-specifier-resolution=node manage-data.import.ts
import assert from 'assert';
const MAX_ROWS_TO_IMPORT = 50;  // limit vectorization calls

// ================================
// ===== INSTANTIATION-COMMON =====
// ================================

// START JSON streaming  // START CSV streaming
import weaviate from 'weaviate-client';
import fs from 'fs';
// END JSON streaming  // END CSV streaming

// START JSON streaming
import parser from 'stream-json';
import StreamArray from 'stream-json/streamers/StreamArray';
import Chain from 'stream-chain';
// END JSON streaming
// START CSV streaming
import csv from 'csv-parser';
// END CSV streaming

// ===== Instantiation, not shown in snippet
const client = await weaviate.connectToWeaviateCloud(
 'WEAVIATE_INSTANCE_URL',  // Replace WEAVIATE_INSTANCE_URL with your instance URL
 {
   authCredentials: new weaviate.ApiKey('api-key'),
   headers: {
     'X-OpenAI-Api-Key': process.env.OPENAI_API_KEY || '',  // Replace with your inference API key
   }
 }
)

// ============================
// ===== Define the class =====
// ============================

const collectionDefinition = {
  name: 'JeopardyQuestion',
  description: 'A Jeopardy! question',
  vectorizers: weaviate.configure.vectorizer.text2VecOpenAI()
};

// Clean slate
try {
  await client.collections.delete('JeopardyQuestion')
  await client.collections.delete('MyCollection')
} catch {
  // ignore error if class doesn't exist
} finally {
  await client.collections.create(collectionDefinition)
}


// ==============================
// ===== Basic batch import =====
// ==============================

{
// START BasicBatchImportExample
let dataObjects = [
  { title: 'Object 1' },
  { title: 'Object 2' },
  { title: 'Object 3' }
]

// highlight-start
const myCollection = client.collections.get('MyCollection')
const response = await myCollection.data.insertMany(dataObjects);
// highlight-end

console.log(response);
// END BasicBatchImportExample

// TODO: update all tests
// let result = await client.graphql.aggregate().withClassName(className).withFields('meta { count }').do();
// assert.equal(result.data['Aggregate'][className][0].meta.count, 5);

client.collections.delete('MyCollection')
}

// =======================================
// ===== Batch import with custom ID =====
// =======================================

// START BatchImportWithIDExample
// highlight-start
import { generateUuid5 } from 'weaviate-client';  // requires v1.3.2+
// highlight-end

// END BatchImportWithIDExample


{
// START BatchImportWithIDExample
let dataObjects = [
  {
    properties: { title: 'Object 1' },
    // highlight-start
    id: generateUuid5('MyCollection', 'Object 1'), // generate uuid from collection name and the unique value(s)
    // highlight-end
  },
  {
    properties: { title: 'Object 2' },
    // highlight-start
    id: generateUuid5('MyCollection', 'Object 2'), // generate uuid from collection name and the unique value(s)
    // highlight-end
  },
  // ...
]

const myCollection = client.collections.get('MyCollection')
await myCollection.data.insertMany(dataObject)
// END BatchImportWithIDExample

// result = await client.graphql.aggregate().withClassName(className).withFields('meta { count }').do();
// assert.equal(result.data['Aggregate'][className][0].meta.count, 5);
// result = await client.graphql.get().withClassName(className).withFields('title _additional { id }').do();
// for (const obj of result.data.Get[className])
//   assert.equal(obj['_additional']['id'], generateUuid5(obj['title']));

await client.collections.delete('MyCollection');
}

// ===========================================
// ===== Batch import with custom vector =====
// ===========================================

// START BatchImportWithVectorExample
const myCollection = client.collections.get('MyCollection')

let dataObjects = [
  {
    properties: { title: 'Object 1' },
    // highlight-start
    vectors: Array(100).fill(0.1111), // provide the vector here
    // highlight-end
  },
  {
    properties: { title: 'Object 2' },
    // highlight-start
    vectors: Array(100).fill(0.2222), // provide the vector here
    // highlight-end
  },
  // ...
]

await jeopardy.data.insertMany(dataObjects)
// END BatchImportWithVectorExample

// result = await client.graphql.aggregate().withClassName(className).withFields('meta { count }').do();
// assert.equal(result.data['Aggregate'][className][0].meta.count, 5);
// result = await client.graphql.get().withClassName(className).withFields('_additional { vector }').do();
// for (const obj of result.data.Get[className]) {
//   assert.ok(obj['_additional']['vector'][0] > 0.25);
//   assert.ok(obj['_additional']['vector'][9] <= 0.3);
// }

await client.collections.delete('MyCollection');


// ===========================================
// ===== Batch import with named vectors =====
// ===========================================

// Clean slate
try {
  await client.collections.delete('MyCollection');
} catch (e) {
  // ignore error if class doesn't exist
} finally {
  await client.collections.create({
    name: 'MyCollection',
    vectorizers: [
      weaviate.configure.vectorizer.none({name: 'title'}),
      weaviate.configure.vectorizer.none({name: 'body'}),
    ]
  })
}

{
// START BatchImportWithNamedVectors
const myCollection = client.collections.get("MyCollection")
let dataObjects = [
  {
    properties: { title: 'Object 1' },
    // highlight-start
    vectors: {
      title: Array(100).fill(0.1111), // provide the vector here
      body: Array(100).fill(0.9999),  // provide the vector here
    }
    // highlight-end
  },
  {
    properties: { title: 'Object 2' },
    // highlight-start
    vectors: {
      title: Array(100).fill(0.2222), // provide the vector here
      body: Array(100).fill(0.8888),  // provide the vector here
    }
    // highlight-end
  },
  // ...
]

await myCollection.data.insertMany(dataObjects)
}
// END BatchImportWithNamedVectors

// Aggregate not working with named vectors as of 2024-02-28
// result = await client.graphql.aggregate().withClassName(className).withFields('meta { count }').do();
// assert.equal(result.data['Aggregate'][className][0].meta.count, 5);

// result = await client.graphql.get().withClassName(className).withFields('_additional { vectors { title body } }').do();
// for (const obj of result.data.Get[className]) {
//   assert.ok(obj['_additional']['vectors']['title']);
//   assert.ok(obj['_additional']['vectors']['body']);
// }

await client.collections.delete('MyCollection');

// ============================
// ===== Streaming import =====
// ============================

// START JSON streaming  // START CSV streaming

let batcher = client.batch.objectsBatcher();
let counter = 0;
const batchSize = 20;

async function addObject(obj: object): Promise<void> {
  const properties = {
    question: obj['Question'],
    answer: obj['Answer'],
  };
  // Add the object to the batch queue
  batcher = batcher.withObject({
    class: 'JeopardyQuestion',
    properties,
    // If you Bring Your Own Vectors, add the `vector` parameter here
    // vector: JSON.parse(obj['Vector']),
  });
  counter++;

  // When the batch counter reaches batchSize, push the objects to Weaviate
  if (counter % batchSize === 0) {
    // Flush the batch queue and restart it
    const response = await batcher.do();
    batcher = client.batch.objectsBatcher();

    // Handle errors
    for (const r of response)
      if (r.result.errors)
        throw r.result.errors;

    console.log(`Imported ${counter} articles...`);
  }
}
// END JSON streaming  // END CSV streaming

// START JSON streaming
async function importJson(filePath) {
  const pipeline = new Chain([
    fs.createReadStream(filePath),
    parser(),
    new StreamArray(),
  ]);

  for await (const { value } of pipeline) {
    await addObject(value);
    // END JSON streaming
    if (counter >= MAX_ROWS_TO_IMPORT)
      return;
    // START JSON streaming
  }
}

await importJson('jeopardy_1k.json');
// END JSON streaming

// START CSV streaming
async function importCSV(filePath) {
  const stream = fs.createReadStream(filePath).pipe(csv());

  for await (const row of stream) {
    await addObject(row);
    // END CSV streaming
    if (counter >= MAX_ROWS_TO_IMPORT * 2)  // CSV import runs after JSON
      return;
    // START CSV streaming
  }
}

await importCSV('jeopardy_1k.csv');
// END CSV streaming

// START JSON streaming  // START CSV streaming

// Flush any remaining objects
if (batcher.payload().objects.length > 0)
  await batcher.do();

console.log(`Finished importing ${counter} articles.`);
// END JSON streaming  // END CSV streaming

// Test
// result = await client.graphql.aggregate().withClassName('JeopardyQuestion').withFields('meta { count }').do();
// assert.deepEqual(result.data['Aggregate']['JeopardyQuestion'], [{ meta: { count: MAX_ROWS_TO_IMPORT * 2 } }]);
// const tokyoRose = await client.graphql.get().withClassName('JeopardyQuestion').withWhere({
//   path: ['answer'],
//   operator: 'Equal',
//   valueText: 'Tokyo Rose',
// }).withFields('answer').do();
// assert.equal(tokyoRose.data.Get['JeopardyQuestion'].length, 2);
// End test

// ===========================
// ===== Batch with gRPC =====
// ===========================
// START BatchGRPC
const questions = client.collections.get("CollectionName")

const batchSize = 1000; // define your batch size

async function insertBatch() {
  try {
    await questions.data.insertMany(dataBatch);
    console.log('Batch inserted successfully');
  } catch (error) {
    console.error('Error inserting batch:', error);
  }
}

async function batchInsert() {
  for (let i = 0; i < dataArray.length; i += batchSize) {
    const batch = dataArray.slice(i, i + batchSize);
    await insertBatch(batch);
  }
}

const dataObject = [...]; // your data
await batchInsert(dataObject);
// START BatchGRPC
